frontend/src/metabase/common/components/EntityPicker/components/EntityPickerModal/EntityPickerModal.tsx

```diff
@@ -119,53 +135,120 @@ export function EntityPickerModal<
+    const hasSearchTab = !!searchQuery;
+    // This is to prevent different tab being initially open and then flickering back
+    // to recents tab once recents have loaded (due to computeInitialTab)
+    const shouldOptimisticallyAddRecentsTabWhileLoading =
+      defaultToRecentTab && isLoadingRecentItems;
+
+    if (hasRecentsTab || shouldOptimisticallyAddRecentsTabWhileLoading) {
+      computedTabs.push({
+        id: RECENTS_TAB_ID,
+        model: null,
+        folderModels: [],
+        displayName: t`Recents`,
+        icon: "clock",
+        render: ({ onItemSelect }) => (
+          <RecentsTab
+            isLoading={isLoadingRecentItems}
+            recentItems={filteredRecents}
+            onItemSelect={onItemSelect}
+            selectedItem={selectedItem}
+          />
+        ),
+      });
+    }
+
+    computedTabs.push(...passedTabs);
+
+    if (hasSearchTab) {
+      computedTabs.push({
+        id: SEARCH_TAB_ID,
+        model: null,
+        folderModels: [],
+        displayName: getSearchTabText(searchResults, searchQuery),
+        icon: "search",
+        render: ({ onItemSelect }) => (
+          <EntityPickerSearchResults
+            searchResults={searchResults}
+            onItemSelect={onItemSelect}
+            selectedItem={selectedItem}
+          />
+        ),
+      });
+    }
+
+    return computedTabs;
+  }, [
+    defaultToRecentTab,
+    filteredRecents,
+    hydratedOptions.hasRecents,
+    isLoadingRecentItems,
+    passedTabs,
+    searchQuery,
+    searchResults,
+    selectedItem,
+  ]);
 
-  const tabs: EntityTab<Model | "recents">[] = useMemo(
-    () =>
-      hydratedOptions.hasRecents && filteredRecents.length > 0
-        ? [
-            {
-              model: "recents",
-              displayName: t`Recents`,
-              icon: "clock",
-              element: (
-                <RecentsTab
-                  isLoading={isLoadingRecentItems}
-                  recentItems={filteredRecents}
-                  onItemSelect={onItemSelect}
-                  selectedItem={selectedItem}
-                />
-              ),
-            },
-            ...passedTabs,
-          ]
-        : passedTabs,
-    [
-      selectedItem,
-      onItemSelect,
-      passedTabs,
-      isLoadingRecentItems,
-      hydratedOptions.hasRecents,
-      filteredRecents,
-    ],
+  const hasTabs = tabs.length > 1;
+  const initialTabId = useMemo(
+    () => computeInitialTabId({ initialValue, tabs, defaultToRecentTab }),
+    [initialValue, tabs, defaultToRecentTab],
   );
+  const [selectedTabId, setSelectedTabId] =
+    useState<EntityPickerTabId>(initialTabId);
+  // we don't want to show bonus actions on recents or search tabs
+  const showActionButtons = ![SEARCH_TAB_ID, RECENTS_TAB_ID].includes(
+    selectedTabId,
+  );
+  const [tabFolderState, setTabFolderState] = useState<TabFolderState>({});
+  const _selectedFolder = tabFolderState[selectedTabId]; // TODO: use me
```

Comment:
These 2 lines are the main goal of this PR.

`EntityPickerModal` will now know which folder is selected in a given tab. We will use this folder to narrow search results in the search tab only to that folder.

More things need to happen before this value will be reliable - we will need to make sure that every tab automatically preselects the first available folder (it already works like that e.g. for questions tab where root collection is automatically selected) and calls `onItemSelect` for that automatic preselection.
This will be done in separate PRs.

----

`_selectedFolder` will be used in other PRs, this TODO has been left intentionally here.